File: checkout-blocked-multi-location-cart/python/detect_multi_location_cart.py
"""Detect Medusa v2 carts stuck at checkout because their items are stocked
across two or more stock locations, a known upstream bug (medusajs/medusa#10561).

The confirm-inventory preparation step, prepare-confirm-inventory-input.ts, merges
every line item's valid stock locations into one flattened list instead of keeping
each item's valid locations scoped to itself. The reserve-inventory step then picks
the first location in that merged list and tries to reserve every item there, so an
item only stocked at a different location fails to reserve, even though the channel
has enough total stock. This lists a cart's items, computes each item's own valid
locations from real per-location stock, and flags the cart when no single location
covers every item, though each item has stock somewhere. Auto-repair is unsafe, since
Medusa v2 has no supported endpoint to force per-item reservation at cart completion,
so the only write here is an optional, DRY_RUN-guarded manual reservation per item at
its own correct location, meant as a one-off mitigation while you upgrade past the bug.

Guide: https://www.allanninal.dev/medusa/checkout-blocked-multi-location-cart/
"""
import os
import logging
import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("detect_multi_location_cart")

BACKEND_URL = os.environ["MEDUSA_BACKEND_URL"]
ADMIN_EMAIL = os.environ["MEDUSA_ADMIN_EMAIL"]
ADMIN_PASSWORD = os.environ["MEDUSA_ADMIN_PASSWORD"]
CART_ID = os.environ.get("CART_ID", "").strip() or None
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"


def get_admin_token():
    r = requests.post(
        f"{BACKEND_URL}/auth/user/emailpass",
        json={"email": ADMIN_EMAIL, "password": ADMIN_PASSWORD},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()["token"]


def get_channel_location_ids(token, sales_channel_id):
    r = requests.get(
        f"{BACKEND_URL}/admin/sales-channels/{sales_channel_id}",
        headers={"Authorization": f"Bearer {token}"},
        params={"fields": "id,*stock_locations"},
        timeout=30,
    )
    r.raise_for_status()
    locations = r.json()["sales_channel"]["stock_locations"]
    return [loc["id"] for loc in locations]


def get_cart(token, cart_id):
    r = requests.get(
        f"{BACKEND_URL}/store/carts/{cart_id}",
        headers={"Authorization": f"Bearer {token}"},
        params={"fields": "id,sales_channel_id,*items,items.variant"},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()["cart"]


def get_variant_inventory_item_id(token, product_id, variant_id):
    r = requests.get(
        f"{BACKEND_URL}/admin/products/{product_id}/variants/{variant_id}",
        headers={"Authorization": f"Bearer {token}"},
        params={"fields": "id,*inventory_items,inventory_items.inventory.id"},
        timeout=30,
    )
    r.raise_for_status()
    items = r.json()["variant"]["inventory_items"]
    return items[0]["inventory"]["id"] if items else None


def get_location_levels(token, inventory_item_id):
    r = requests.get(
        f"{BACKEND_URL}/admin/inventory-items/{inventory_item_id}/location-levels",
        headers={"Authorization": f"Bearer {token}"},
        params={"fields": "location_id,stocked_quantity,reserved_quantity"},
        timeout=30,
    )
    r.raise_for_status()
    levels = r.json()["inventory_levels"]
    return [
        {
            "locationId": lvl["location_id"],
            "stockedQuantity": lvl["stocked_quantity"],
            "reservedQuantity": lvl["reserved_quantity"],
        }
        for lvl in levels
    ]


def has_reservation(token, line_item_id):
    r = requests.get(
        f"{BACKEND_URL}/admin/reservations",
        headers={"Authorization": f"Bearer {token}"},
        params={"line_item_id": line_item_id},
        timeout=30,
    )
    r.raise_for_status()
    return len(r.json().get("reservations", [])) > 0


def create_reservation(token, line_item_id, inventory_item_id, location_id, quantity):
    r = requests.post(
        f"{BACKEND_URL}/admin/reservations",
        headers={"Authorization": f"Bearer {token}", "Content-Type": "application/json"},
        json={
            "line_item_id": line_item_id,
            "inventory_item_id": inventory_item_id,
            "location_id": location_id,
            "quantity": quantity,
        },
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def resolve_item_locations(items, levels_by_inventory_item, channel_location_ids):
    """Pure decision function. No I/O.

    items: [{"lineItemId": str, "inventoryItemId": str, "requiredQty": int}, ...]
    levels_by_inventory_item: {inventoryItemId: [{"locationId": str, "stockedQuantity": int,
                                                   "reservedQuantity": int}, ...]}
    channel_location_ids: [str, ...]

    Returns [{"lineItemId": str, "validLocationIds": [str, ...]}, ...], one entry per
    item, in the same order, where validLocationIds is every channel-linked location
    that has enough available stock (stockedQuantity - reservedQuantity) for that item.
    """
    channel_set = set(channel_location_ids)
    results = []
    for item in items:
        levels = levels_by_inventory_item.get(item["inventoryItemId"], [])
        valid_location_ids = [
            lvl["locationId"]
            for lvl in levels
            if lvl["locationId"] in channel_set
            and (lvl["stockedQuantity"] - lvl["reservedQuantity"]) >= item["requiredQty"]
        ]
        results.append({"lineItemId": item["lineItemId"], "validLocationIds": valid_location_ids})
    return results


def is_affected_cart(item_locations):
    """Pure decision function. No I/O.

    A cart is a stuck-at-reservation candidate when every item has at least one valid
    location of its own, but no single location is valid for every item at once, the
    signature of medusajs/medusa#10561. An empty cart, or a cart where some item has
    zero valid locations (a real out-of-stock case, not this bug), is not affected.
    """
    if not item_locations:
        return False
    if any(len(entry["validLocationIds"]) == 0 for entry in item_locations):
        return False
    shared = set(item_locations[0]["validLocationIds"])
    for entry in item_locations[1:]:
        shared &= set(entry["validLocationIds"])
    return len(shared) == 0


def run():
    if not CART_ID:
        raise RuntimeError("Set CART_ID to the cart you want to check.")

    token = get_admin_token()
    cart = get_cart(token, CART_ID)
    channel_location_ids = get_channel_location_ids(token, cart["sales_channel_id"])

    items = []
    inventory_item_by_line_item = {}
    for line_item in cart["items"]:
        variant = line_item["variant"]
        inventory_item_id = get_variant_inventory_item_id(token, variant["product_id"], variant["id"])
        if not inventory_item_id:
            continue
        items.append({
            "lineItemId": line_item["id"],
            "inventoryItemId": inventory_item_id,
            "requiredQty": line_item["quantity"],
        })
        inventory_item_by_line_item[line_item["id"]] = inventory_item_id

    levels_by_inventory_item = {
        item["inventoryItemId"]: get_location_levels(token, item["inventoryItemId"])
        for item in items
    }

    item_locations = resolve_item_locations(items, levels_by_inventory_item, channel_location_ids)
    affected = is_affected_cart(item_locations)

    if not affected:
        log.info("Cart %s: not affected. Items share a common valid location, or one has none at all.", CART_ID)
        return

    unreserved = [entry for entry in item_locations if not has_reservation(token, entry["lineItemId"])]
    log.warning(
        "Cart %s is affected by medusajs/medusa#10561: no shared location across items, "
        "%d item(s) missing a reservation.",
        CART_ID, len(unreserved),
    )
    for entry in item_locations:
        log.info("  line item %s valid locations: %s", entry["lineItemId"], entry["validLocationIds"])

    for entry in unreserved:
        location_id = entry["validLocationIds"][0]
        inventory_item_id = inventory_item_by_line_item[entry["lineItemId"]]
        required_qty = next(i["requiredQty"] for i in items if i["lineItemId"] == entry["lineItemId"])
        payload = {
            "line_item_id": entry["lineItemId"],
            "inventory_item_id": inventory_item_id,
            "location_id": location_id,
            "quantity": required_qty,
        }
        log.info("%s reservation: %s", "Would create" if DRY_RUN else "Creating", payload)
        if not DRY_RUN:
            create_reservation(token, **{
                "line_item_id": payload["line_item_id"],
                "inventory_item_id": payload["inventory_item_id"],
                "location_id": payload["location_id"],
                "quantity": payload["quantity"],
            })

    log.info("Done. %d item(s) %s a manual reservation.", len(unreserved), "would need" if DRY_RUN else "given")


if __name__ == "__main__":
    run()

File: checkout-blocked-multi-location-cart/node/detect-multi-location-cart.js
/**
 * Detect Medusa v2 carts stuck at checkout because their items are stocked
 * across two or more stock locations, a known upstream bug (medusajs/medusa#10561).
 *
 * The confirm-inventory preparation step, prepare-confirm-inventory-input.ts, merges
 * every line item's valid stock locations into one flattened list instead of keeping
 * each item's valid locations scoped to itself. The reserve-inventory step then picks
 * the first location in that merged list and tries to reserve every item there, so an
 * item only stocked at a different location fails to reserve, even though the channel
 * has enough total stock. This lists a cart's items, computes each item's own valid
 * locations from real per-location stock, and flags the cart when no single location
 * covers every item, though each item has stock somewhere. Auto-repair is unsafe, since
 * Medusa v2 has no supported endpoint to force per-item reservation at cart completion,
 * so the only write here is an optional, DRY_RUN-guarded manual reservation per item at
 * its own correct location, meant as a one-off mitigation while you upgrade past the bug.
 *
 * Guide: https://www.allanninal.dev/medusa/checkout-blocked-multi-location-cart/
 */
import { pathToFileURL } from "node:url";

const BACKEND_URL = process.env.MEDUSA_BACKEND_URL || "http://localhost:9000";
const ADMIN_EMAIL = process.env.MEDUSA_ADMIN_EMAIL || "admin@example.com";
const ADMIN_PASSWORD = process.env.MEDUSA_ADMIN_PASSWORD || "supersecret";
const CART_ID = (process.env.CART_ID || "").trim() || null;
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

/**
 * Pure decision function. No I/O.
 *
 * @param {{ lineItemId: string, inventoryItemId: string, requiredQty: number }[]} items
 * @param {Record<string, { locationId: string, stockedQuantity: number, reservedQuantity: number }[]>} levelsByInventoryItem
 * @param {string[]} channelLocationIds
 * @returns {{ lineItemId: string, validLocationIds: string[] }[]}
 */
export function resolveItemLocations(items, levelsByInventoryItem, channelLocationIds) {
  const channelSet = new Set(channelLocationIds);
  return items.map((item) => {
    const levels = levelsByInventoryItem[item.inventoryItemId] || [];
    const validLocationIds = levels
      .filter(
        (lvl) =>
          channelSet.has(lvl.locationId) &&
          lvl.stockedQuantity - lvl.reservedQuantity >= item.requiredQty
      )
      .map((lvl) => lvl.locationId);
    return { lineItemId: item.lineItemId, validLocationIds };
  });
}

/**
 * Pure decision function. No I/O.
 *
 * A cart is a stuck-at-reservation candidate when every item has at least one valid
 * location of its own, but no single location is valid for every item at once, the
 * signature of medusajs/medusa#10561.
 *
 * @param {{ lineItemId: string, validLocationIds: string[] }[]} itemLocations
 * @returns {boolean}
 */
export function isAffectedCart(itemLocations) {
  if (!itemLocations.length) return false;
  if (itemLocations.some((entry) => entry.validLocationIds.length === 0)) return false;
  const shared = itemLocations.reduce(
    (acc, entry) => new Set([...acc].filter((id) => entry.validLocationIds.includes(id))),
    new Set(itemLocations[0].validLocationIds)
  );
  return shared.size === 0;
}

async function getSdk() {
  const { default: Medusa } = await import("@medusajs/js-sdk");
  const sdk = new Medusa({ baseUrl: BACKEND_URL, auth: { type: "jwt" } });
  await sdk.auth.login("user", "emailpass", { email: ADMIN_EMAIL, password: ADMIN_PASSWORD });
  return sdk;
}

async function getChannelLocationIds(sdk, salesChannelId) {
  const { sales_channel } = await sdk.admin.salesChannel.retrieve(salesChannelId, {
    fields: "id,*stock_locations",
  });
  return sales_channel.stock_locations.map((loc) => loc.id);
}

async function getCart(sdk, cartId) {
  const { cart } = await sdk.store.cart.retrieve(cartId, {
    fields: "id,sales_channel_id,*items,items.variant",
  });
  return cart;
}

async function getVariantInventoryItemId(sdk, productId, variantId) {
  const { variant } = await sdk.admin.product.retrieveVariant(productId, variantId, {
    fields: "id,*inventory_items,inventory_items.inventory.id",
  });
  const items = variant.inventory_items;
  return items.length ? items[0].inventory.id : null;
}

async function getLocationLevels(sdk, inventoryItemId) {
  const { inventory_levels } = await sdk.admin.inventoryItem.listLocationLevels(inventoryItemId, {
    fields: "location_id,stocked_quantity,reserved_quantity",
  });
  return inventory_levels.map((lvl) => ({
    locationId: lvl.location_id,
    stockedQuantity: lvl.stocked_quantity,
    reservedQuantity: lvl.reserved_quantity,
  }));
}

async function hasReservation(sdk, lineItemId) {
  const { reservations } = await sdk.admin.reservation.list({ line_item_id: lineItemId });
  return reservations.length > 0;
}

async function createReservation(sdk, payload) {
  return sdk.admin.reservation.create(payload);
}

export async function run() {
  if (!CART_ID) throw new Error("Set CART_ID to the cart you want to check.");

  const sdk = await getSdk();
  const cart = await getCart(sdk, CART_ID);
  const channelLocationIds = await getChannelLocationIds(sdk, cart.sales_channel_id);

  const items = [];
  const inventoryItemByLineItem = new Map();
  for (const lineItem of cart.items) {
    const variant = lineItem.variant;
    const inventoryItemId = await getVariantInventoryItemId(sdk, variant.product_id, variant.id);
    if (!inventoryItemId) continue;
    items.push({ lineItemId: lineItem.id, inventoryItemId, requiredQty: lineItem.quantity });
    inventoryItemByLineItem.set(lineItem.id, inventoryItemId);
  }

  const levelsByInventoryItem = {};
  for (const item of items) {
    levelsByInventoryItem[item.inventoryItemId] = await getLocationLevels(sdk, item.inventoryItemId);
  }

  const itemLocations = resolveItemLocations(items, levelsByInventoryItem, channelLocationIds);
  const affected = isAffectedCart(itemLocations);

  if (!affected) {
    console.log(`Cart ${CART_ID}: not affected. Items share a common valid location, or one has none at all.`);
    return;
  }

  const unreserved = [];
  for (const entry of itemLocations) {
    if (!(await hasReservation(sdk, entry.lineItemId))) unreserved.push(entry);
  }

  console.warn(
    `Cart ${CART_ID} is affected by medusajs/medusa#10561: no shared location across items, ${unreserved.length} item(s) missing a reservation.`
  );
  for (const entry of itemLocations) {
    console.log(`  line item ${entry.lineItemId} valid locations: ${JSON.stringify(entry.validLocationIds)}`);
  }

  for (const entry of unreserved) {
    const locationId = entry.validLocationIds[0];
    const inventoryItemId = inventoryItemByLineItem.get(entry.lineItemId);
    const requiredQty = items.find((i) => i.lineItemId === entry.lineItemId).requiredQty;
    const payload = {
      line_item_id: entry.lineItemId,
      inventory_item_id: inventoryItemId,
      location_id: locationId,
      quantity: requiredQty,
    };
    console.log(`${DRY_RUN ? "Would create" : "Creating"} reservation: ${JSON.stringify(payload)}`);
    if (!DRY_RUN) await createReservation(sdk, payload);
  }

  console.log(`Done. ${unreserved.length} item(s) ${DRY_RUN ? "would need" : "given"} a manual reservation.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}
